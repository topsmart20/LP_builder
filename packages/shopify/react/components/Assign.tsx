import React from 'react';
import { Builder, BuilderStore } from '@builder.io/react';

interface AssignBlockProps {
  expression?: string;
  builderState?: BuilderStore;
}

export class AssignBlock extends React.Component<AssignBlockProps> {
  constructor(props: AssignBlockProps) {
    super(props);

    const { expression, builderState } = props;

    if (expression && builderState) {
      builderState.state.shopify?.liquid?.assign(expression);
    }
  }
}

Builder.registerComponent(AssignBlock, {
  name: 'Shopify:Assign',
  hideFromInsertMenu: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
      // TODO: type: 'code', language: 'javascript'
    },
  ],
});
