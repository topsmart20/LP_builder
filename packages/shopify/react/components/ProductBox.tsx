/** @jsx jsx */
import { jsx, css, keyframes } from '@emotion/core';

import React, { useEffect, useState } from 'react';
import { Builder, BuilderElement } from '@builder.io/sdk';
import { BuilderStore, BuilderStoreContext, BuilderBlocks } from '@builder.io/react';
import { modifyProduct } from '../functions/modify-product';

interface ProductBoxProps {
  product?: string | number;
  /** @todo implement this */
  variant?: string | number;
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
}

const spin = keyframes`
  0% {
    transform:rotate(0)
  }
  100% {
    transform:rotate(360deg)
  }
`;

function LoadingSpinner() {
  return (
    <div
      css={css`
        width: 4em;
        height: 4em;
        border-radius: 50%;
        position: relative;
        margin: 6rem auto;
        font-size: 1rem;
        text-indent: -9999em;
        border-top: 0.2em solid rgba(131, 132, 137, 0.2);
        border-right: 0.2em solid rgba(131, 132, 137, 0.2);
        border-bottom: 0.2em solid rgba(131, 132, 137, 0.2);
        border-left: 0.2em solid #454749;
        transform: translateZ(0);
        animation: 1.1s linear infinite ${spin};
      `}
    />
  );
}

export function ProductBox(props: ProductBoxProps) {
  const [productInfo, setProductInfo] = useState(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    updateProduct();
  }, [props.product]);

  function updateProduct() {
    setLoading(true);
    const { context } = props.builderState!;
    let productId = props.product || '';

    if (productId) {
      fetch(`https://builder.io/api/v1/shopify/products/${productId}.json?apiKey=${context.apiKey}`)
        .then(res => res.json())
        .then(data => {
          const { product } = data;
          const modifiedProduct = modifyProduct(product);
          setProductInfo(modifiedProduct);
          setLoading(false);
        })
        .catch(err => {
          console.error('Error fetching Shopify product', err);
          setLoading(false);
        });
    } else {
      setProductInfo(null);
    }
  }

  return loading && !productInfo ? (
    <LoadingSpinner />
  ) : (
    <BuilderStoreContext.Consumer>
      {store => (
        <BuilderStoreContext.Provider
          value={{
            ...store,
            state: {
              ...store.state,
              // TODO: communicate up this up to DataEditor to populate bindings dropdowns (how?)
              productInfo,
            },
          }}
        >
          <BuilderBlocks
            child
            parentElementId={props.builderBlock?.id}
            blocks={props.builderBlock?.children}
          />
        </BuilderStoreContext.Provider>
      )}
    </BuilderStoreContext.Consumer>
  );
}

Builder.registerComponent(ProductBox, {
  name: 'Shopify:ProductBox',
  canHaveChildren: true,
  hideFromInsertMenu: true,
  inputs: [
    {
      name: 'product',
      type: 'ShopifyProduct',
    },
  ],
});
