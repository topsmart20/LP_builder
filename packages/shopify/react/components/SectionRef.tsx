import React, { useState, useEffect } from 'react';
import { Builder } from '@builder.io/react';

interface SectionRefProps {
  class?: string;
  section?: string;
}
const cache: { [key: string]: string } = {};

export const SectionRef = (props: SectionRefProps) => {
  const [html, setHtml] = useState<string | null>(null);
  const sectionName = props.section?.split('/')[1].split('.')[0]!;

  useEffect(() => {
    if (cache[sectionName]) {
      return;
    }
    // Convert `sections/foo.liquid` -> `foo`
    if (!sectionName) {
      return;
    }
    if (cache[sectionName]) {
      setHtml(cache[sectionName]);
      return;
    }
    // This may look strange, but it is how Shopify's section rendering API
    // works https://shopify.dev/docs/themes/sections/section-rendering-api
    fetch(`${location.search ? '&' : '?'}section_id=${sectionName}`)
      .then(res => res.text())
      .then(text => {
        cache[sectionName] = text;
        setHtml(text);
      });
  }, [sectionName]);

  return (
    <div
      className={`shopify-section ${props.class}`}
      dangerouslySetInnerHTML={{ __html: html || cache[sectionName] || '' }}
    />
  );
};

Builder.registerComponent(SectionRef, {
  name: 'Shopify:SectionRef',
  hideFromInsertMenu: true,
});
