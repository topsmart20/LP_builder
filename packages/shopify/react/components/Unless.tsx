import * as React from 'react';
import builder, {
  Builder,
  BuilderStore,
  onChange,
  withBuilder,
  BuilderElement,
  stringToFunction,
  BuilderBlocks,
} from '@builder.io/react';

interface UnlessBlockProps {
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
  unlessBlocks?: BuilderElement[];
  elseBlocks?: BuilderElement[];
  expression?: string;
}

export class UnlessBlock extends React.Component<UnlessBlockProps> {
  private getMatchingBlocks() {
    const { expression, unlessBlocks, elseBlocks } = this.props;
    if (!expression) {
      return unlessBlocks;
    }
    const fn = stringToFunction(expression);
    const result = fn(
      this.props.builderState && this.props.builderState.state,
      null,
      this.props.builderBlock,
      builder,
      null,
      this.props.builderState!.update,
      Builder,
      this.props.builderState?.context
    );
    return result ? elseBlocks : unlessBlocks;
  }

  render() {
    const { unlessBlocks } = this.props;
    const blocks = this.getMatchingBlocks();
    if (!blocks) {
      return null;
    }
    const isUnless = blocks === unlessBlocks;

    return (
      <BuilderBlocks
        child
        parentElementId={this.props.builderBlock && this.props.builderBlock.id}
        blocks={blocks}
        dataPath={`component.options.${isUnless ? 'unlessBlocks' : 'elseBlocks'}`}
      />
    );
  }
}

const defaultBlock = {
  '@type': '@builder.io/sdk:Element',
  responsiveStyles: {
    large: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'stretch',
      flexShrink: '0',
      position: 'relative',
      marginTop: '30px',
      textAlign: 'center',
      lineHeight: 'normal',
      height: 'auto',
    },
  },
  component: {
    name: 'Text',
    options: {
      text: '<p>Enter some text...</p>',
    },
  },
};

withBuilder(UnlessBlock, {
  name: 'Shopify:Unless',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
    {
      name: 'unlessBlocks',
      type: 'array',
      hideFromUI: true,
      defaultValue: [defaultBlock],
    },
    {
      name: 'elseBlocks',
      type: 'array',
      hideFromUI: true,
      defaultValue: [defaultBlock],
    },
  ],
});
