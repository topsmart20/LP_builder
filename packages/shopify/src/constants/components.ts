import { BuilderElement } from "@builder.io/sdk";
import { Options } from "../interfaces/options";

export interface ComponentInfo {
  component: (block: BuilderElement, options: Options, attributes?: string) => string;
  noWrap?: boolean;
  name: string;
}

export function component(info: ComponentInfo) {
  components[info.name] = info;
  return info.component;
}

export function getComponentInfo(name: string) {
  return components[name];
}

export function getComponent(name: string) {
  const info = getComponentInfo(name)
  return info && info.component;
}

export const components: { [key: string]: ComponentInfo | undefined } = {};
