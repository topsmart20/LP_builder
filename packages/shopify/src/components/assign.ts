import { BuilderElement } from '@builder.io/sdk';
import { Options } from '../interfaces/options';
import { component } from '../constants/components';

export const Assign = component({
  name: 'Shopify:Assign',
  noWrap: true,
  component: (block: BuilderElement, renderOptions: Options) => {
    const { options } = block.component!;
    const content = options.expression;

    return `{% assign ${content} %}`;
  },
});
