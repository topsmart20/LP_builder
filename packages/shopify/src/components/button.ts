import { component } from '../constants/components';

export const Button = component({
  name: 'Core:Button',
  noWrap: true,
  component: (block, renderOptions, attributes) => {
    const { options } = block.component!;
    const tag = options.link ? 'a' : 'span';

    return `
      <${tag}${options.openLinkInNewTab ? ' target="_blank"' : ''} ${attributes || ''}>
        ${options.text || ''}
      </${tag}>
    `;
  },
});
