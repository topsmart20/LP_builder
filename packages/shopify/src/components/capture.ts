import { BuilderElement } from '@builder.io/sdk';
import { Options } from '../interfaces/options';
import { component } from '../constants/components';

export const Capture = component({
  name: 'Shopify:Capture',
  noWrap: true,
  component: (block: BuilderElement, renderOptions: Options) => {
    const { options } = block.component!;
    const { expression, variableName } = options;

    return `{% capture ${variableName} %}${expression}{% endcapture %}`;
  },
});
