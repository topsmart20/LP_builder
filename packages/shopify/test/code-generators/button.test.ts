import { Button } from '../../src/components/button';
import { el } from '../modules/helpers';

const getButtonCode = (options: any): string =>
  Button(
    el({
      component: {
        options,
        name: 'Core:Button',
      },
    }),
    {}
  );

describe('Button code generator', () => {
  test('without link', () => {
    const code = getButtonCode({
      text: 'Click me',
    });

    expect(code).toMatchSnapshot();
  });

  test('with link', () => {
    const code = getButtonCode({
      text: 'Click me',
      link: true,
    });

    expect(code).toMatchSnapshot();
  });

  test('with link in new tab, no text', () => {
    const code = getButtonCode({
      link: true,
      openLinkInNewTab: true,
    });

    expect(code).toMatchSnapshot();
  });
});
