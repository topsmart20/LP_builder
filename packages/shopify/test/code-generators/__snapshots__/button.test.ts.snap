// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button code generator with link 1`] = `
"
      <a >
        Click me
      </a>
    "
`;

exports[`Button code generator with link in new tab, no text 1`] = `
"
      <a target=\\"_blank\\" >
        
      </a>
    "
`;

exports[`Button code generator without link 1`] = `
"
      <span >
        Click me
      </span>
    "
`;
